### ISSUE TITLE
Please make an option to set compatibility mode automatically when IWAD loaded.

### ISSUE THREAD

**Issue by skyrain128** [coelckers/prboom-plus#264]
Please make an option to set compatibility mode automatically when IWAD loaded.
(ex: If I selected Ultimate Doom IWAD, set compatibility mode automatically to Ultimate Doom IWAD.)

**Comment by fabiangreffrath** [comment]
That would be `-complevel vanilla`, right?

**Comment by skyrain128** [comment]
Yes, but could you make it can be set in in-game's options?

**Comment by fabiangreffrath** [comment]
This could work:
```patch
diff --git a/prboom2/src/d_server.c b/prboom2/src/d_server.c
index 39c2e3cd..a94571f7 100644
--- a/prboom2/src/d_server.c
+++ b/prboom2/src/d_server.c
@@ -320,6 +320,7 @@ void read_config_file(FILE* fp, struct setup_packet_s* sp)
         sp->skill = v-1;
       } else if (!strcmp(def,"default_compatibility_level")) {
         if (verbose) printf("config file sets compatibility_level to %d\n",v);
+	if (v == -2) v = 2; // [FG] -2 => vanilla
 	if (v == -1) v = MAX_COMPATIBILITY_LEVEL-1; //e6y: -1 => maxcompat
         sp->complevel = v;
       } else {
diff --git a/prboom2/src/g_game.c b/prboom2/src/g_game.c
index 99e1bba4..6ff194c9 100644
--- a/prboom2/src/g_game.c
+++ b/prboom2/src/g_game.c
@@ -2740,6 +2740,8 @@ void G_ReloadDefaults(void)
       if (l >= -1) compatibility_level = l;
     }
   }
+  if (compatibility_level == -2)
+    compatibility_level = G_GetNamedComplevel("vanilla");
   if (compatibility_level == -1)
     compatibility_level = best_compatibility;
 
diff --git a/prboom2/src/m_menu.c b/prboom2/src/m_menu.c
index ebbe4358..7be24ae6 100644
--- a/prboom2/src/m_menu.c
+++ b/prboom2/src/m_menu.c
@@ -3250,6 +3250,7 @@ static const char *gen_skillstrings[] = {
 
 static const char *gen_compstrings[] =
 {
+  "Vanilla",
   "Default",
   "Doom v1.2",
   "Doom v1.666",
@@ -3288,7 +3289,7 @@ setup_menu_t gen_settings2[] = { // General Settings screen2
   {"Maximum number of player corpses", S_NUM|S_PRGWARN, m_null, G_X, G_Y+12*8, {"max_player_corpse"}},
   {"Game speed, percentage of normal", S_NUM|S_PRGWARN, m_null, G_X, G_Y+13*8, {"realtic_clock_rate"}},
   {"Default skill level",              S_CHOICE,        m_null, G_X, G_Y+14*8, {"default_skill"}, 0, 0, NULL, gen_skillstrings},
-  {"Default compatibility level",      S_CHOICE,        m_null, G_X, G_Y+15*8, {"default_compatibility_level"}, 0, 0, NULL, &gen_compstrings[1]},
+  {"Default compatibility level",      S_CHOICE,        m_null, G_X, G_Y+15*8, {"default_compatibility_level"}, 0, 0, NULL, &gen_compstrings[2]},
   {"Show ENDOOM screen",               S_YESNO,         m_null, G_X, G_Y+16*8, {"showendoom"}},
   {"Fullscreen menu background",       S_YESNO, m_null, G_X, G_Y + 17*8, {"menu_background"}},
 #ifdef USE_WINDOWS_LAUNCHER
diff --git a/prboom2/src/m_misc.c b/prboom2/src/m_misc.c
index 55820344..3eaa3124 100644
--- a/prboom2/src/m_misc.c
+++ b/prboom2/src/m_misc.c
@@ -301,7 +301,7 @@ default_t defaults[] =
   
   {"Misc settings",{NULL},{0},UL,UL,def_none,ss_none},
   {"default_compatibility_level",{(int*)&default_compatibility_level},
-   {-1},-1,MAX_COMPATIBILITY_LEVEL-1,
+   {-1},-2,MAX_COMPATIBILITY_LEVEL-1,
    def_int,ss_none}, // compatibility level" - CPhipps
   {"realtic_clock_rate",{&realtic_clock_rate},{100},0,UL,
    def_int,ss_none}, // percentage of normal speed (35 fps) realtic clock runs at
```
The code in `d_server.c` is a bit hacky at the moment.

**Comment by fabiangreffrath** [comment]
I still don't see that much advantage over just setting "Default compatibility level" to "Doom/2 v1.9" that would justify a change as invasive as this.

**Comment by JadingTsunami** [comment]
> I still don't see that much advantage over just setting "Default compatibility level" to "Doom/2 v1.9" that would justify a change as invasive as this.

I tend to agree. I am not sure this change has widespread appeal for users.

**Comment by RamonUnch** [comment]
I think this makes sense for demo recording. If you want to record vanilla demos and play to different IWADS then you no longer need to specify the -complevel on the command line, except for occasional boom/MBF recordings.
From a user perspective, I think adding this option would actually be beneficial.
Default compatibility level to Doom2/v1.9 is a broken option in my opinion because it breaks ultimate doom and final doom demo recording (and even player's experience in final doom) unless you remember to specify -complevel.

In term of options I like to have the fewest and the most relevant, so i think that in the GUI complevel 2 3 and 4 should not even be available if vanilla is added. If a player wants to record a demo with non  IWAD matching complevel then he can always specify the complevel on the command line.

Of course this is my opinion, but I think that for new users it would be helpful. PrBoom+ has way too many options and if they could be reduced everyone would benefit.

**Comment by JadingTsunami** [comment]
I see the point, and speaking generally I am in favor of smart defaults. I just wonder if this is a job better suited to a separate front-end or GUI-based launcher utility. Trying to predict the user's intent from incomplete information can be tricky.

**Comment by RamonUnch** [comment]
I agree that there is no perfect solutions because the matter is quite complex.
My point is that the current options are already quite misleading for a non expert.
and the idea of -complevel vanilla is indeed to consider that 1.9 1.9u and 1.9f  constitute a *meta port* because it is the original behavior that you would have had with the original exe for each game.
For ports complevel, it is simpler because there is no dependence on the IWAD.

I agree that it makes things internally more complicated and maybe it does not really simplifies the usage.
But I think that here we are not really trying to predict the user intent, because it is clear that no one wants to have complevel 2 with ultimate of final doom outside of obscure tests. Especially if someone set the default compatibility level to 2 for sure he wants to play with the "proper" vanilla compat and and not Doom2/v1.9.

A graphical launcher could indeed handle this, however when recording the prboom+ is directly invoked from the command line, and the ability to have the equivalent of -complevel vanilla set in the cfg, instead of typing it every time, is seducing.
Personally when I play with prboom I always manually specifies the complevel in the command line and I never use the cfg option for this very reason.

I just give my 2c on this, do not feel obliged or anything.

**Comment by fabiangreffrath** [comment]
> Of course this is my opinion, but I think that for new users it would be helpful. PrBoom+ has way too many options and if they could be reduced everyone would benefit.

In general I agree. But providing support for even the most obscure demo formats is this port's explicit mission goal, so this is probably not the right place to reduce the number of available options.

> But I think that here we are not really trying to predict the user intent, because it is clear that no one wants to have complevel 2 with ultimate of final doom outside of obscure tests. Especially if someone set the default compatibility level to 2 for sure he wants to play with the "proper" vanilla compat and and not Doom2/v1.9.

Sorry, but this is plain wrong. There are Final Doom demos out there that have been recorded with a Doom 2 executable. And if someone sets the compatibility level to "Doom2/1.9" and plays The Ultimate Doom with it, I am pretty sure he wants Doom 2 settings, because else he would have set the compatibility level to "Ultimate Doom" which is available right next to it.
